tests/quickcheck.rs

```diff
@@ -0,0 +1,326 @@
+            a.clone() * (b.clone() + c.clone()) == a.clone() * b + a * c
+        }
+    }
+
+    quickcheck! {
+        fn quickcheck_signed_distributive(a: BigInt, b: BigInt, c: BigInt) -> bool {
+            a.clone() * (b.clone() + c.clone()) == a.clone() * b + a * c
+        }
+    }
+
+    quickcheck! {
+        ///Tests that exactly one of a<b a>b a=b is true
+        fn quickcheck_unsigned_ge_le_eq_mut_exclusive(a: BigUint, b: BigUint) -> bool {
+            let gt_lt_eq = vec![a > b, a < b, a == b];
+            gt_lt_eq
+                .iter()
+                .fold(0, |acc, e| if *e { acc + 1 } else { acc })
+                == 1
+        }
+    }
+
+    quickcheck! {
+        ///Tests that exactly one of a<b a>b a=b is true
+        fn quickcheck_signed_ge_le_eq_mut_exclusive(a: BigInt, b: BigInt) -> bool {
+            let gt_lt_eq = vec![a > b, a < b, a == b];
+            gt_lt_eq
+                .iter()
+                .fold(0, |acc, e| if *e { acc + 1 } else { acc })
+                == 1
+        }
+    }
+
+    quickcheck! {
+        fn quickcheck_unsigned_sub_primitive(a: u128, b: u128) -> bool {
+            if b < a {
+                BigUint::from(a-b) == BigUint::from(a) - BigUint::from(b)
+            } else {
+                BigUint::from(b-a) == BigUint::from(b) - BigUint::from(a)
+            }
+        }
+    }
+
+    quickcheck! {
+        fn quickcheck_signed_sub_primitive(a: i128, b: i128) -> bool {
+            if b < a {
+                BigInt::from(a-b) == BigInt::from(a) - BigInt::from(b)
+            } else {
+                BigInt::from(b-a) == BigInt::from(b) - BigInt::from(a)
+            }
+        }
+    }
+
+    quickcheck! {
+        /// Tests correctness of subtraction assuming addition is correct
+        fn quickcheck_unsigned_sub(a: BigUint, b: BigUint) -> bool {
+            if b < a {
+                a.clone() - b.clone() + b == a
+            } else {
+                b.clone() - a.clone() + a == b
+            }
+        }
+    }
+
+    quickcheck! {
+        /// Tests correctness of subtraction assuming addition is correct
+        fn quickcheck_signed_sub(a: BigInt, b: BigInt) -> bool {
+            if b < a {
+                a.clone() - b.clone() + b == a
+            } else {
+                b.clone() - a.clone() + a == b
+            }
+        }
+    }
+
+    quickcheck! {
+        fn quickcheck_unsigned_div_primitive(a: u128, b: u128) -> TestResult {
+            if b == 0 {
+                TestResult::discard()
+            } else {
+                TestResult::from_bool(BigUint::from(a/b) == BigUint::from(a) / BigUint::from(b))
+            }
+        }
+    }
+
+    quickcheck! {
+        fn quickcheck_signed_div_primitive(a: i128, b: i128) -> TestResult {
+            if b == 0 {
+                TestResult::discard()
+            } else {
+                TestResult::from_bool(BigInt::from(a/b) == BigInt::from(a) / BigInt::from(b))
+            }
+        }
+    }
```

Comment:
How about `Rem`, `Shl`, `Shr`, etc.? For example, we had a problem with `Shr` in #1, fixed to match primitives in #8, that would have been nice to discover sooner. I suspect quickcheck would have triggered it with ease. If you're curious, you could try breaking `shr_round_down` to return false and see what happens.

Reply:
There is a significant and annoying problem with testing shift operations (really, everything, but particularly shifts). `quickcheck` currently only generates numbers in `0..100` for unsigned types (I think `-100..100` for signed types). This makes for poor bit shift testing, because the upper bits of the number will always be zero. See https://github.com/BurntSushi/quickcheck/issues/190. I consider this issue blocking for bitshift quickchecks, as trivial tests pass when they would not if the range were the full primitive's range. Then if quickcheck were updated to check the whole range, it would break our tests (unless our tests were written to accommodate changing ranges, but this strategy seems risky). I may submit a PR to quickcheck for this.

For now I will investigate the comment on that issue:
> You can however change the range at which usize numbers are generated by creating a StdGen with a custom size and setting it on QuickCheck.